rust_src/src/dired_unix.rs

```diff
@@ -95,6 +100,270 @@ impl LispObjectExt for LispObject {
     }
 }
 
+struct DirFiles {
+    directory: String,
+    full: bool,
+    match_re: LispObject,
+    nosort: bool,
+    attrs: bool,
+    id_format: LispObject,
+    files: Vec<String>,
+    files_attrs: Vec<LispObject>,
+}
+
+impl DirFiles {
+    fn new(
+        directory: String,
+        full: bool,
+        match_re: LispObject,
+        nosort: bool,
+        attrs: bool,
+        id_format: LispObject,
+    ) -> Self {
+        Self {
+            directory,
+            full,
+            match_re,
+            nosort,
+            attrs,
+            id_format,
+            files: Vec::new(),
+            files_attrs: Vec::new(),
+        }
+    }
+
+    // read_dir() does not return .&.. so bolt them on
+    fn add_dots(&mut self) {
+        let d = String::from(".");
+        let dd = String::from("..");
+        let mut df = d.clone(); // dot final name
+        let mut ddf = dd.clone();
```

Comment:
This has quite a lot of cloning. I don't see that any of the clones get written to, so there should be a way to avoid them.